server/datastore/mysql/migrations/tables/20250410104321_UpdateMacOSSoftwareNames.go

```diff
@@ -13,52 +15,59 @@ func init() {
 	MigrationClient.AddMigration(Up_20250410104321, Down_20250410104321)
 }
 
-// TODO(JVE): add bundleid != null to all constraints
 func Up_20250410104321(tx *sql.Tx) error {
 	titleStmt := `UPDATE software_titles SET name = TRIM( TRAILING '.app' FROM name ) WHERE source = 'apps' AND bundle_identifier IS NOT NULL`
 	_, err := tx.Exec(titleStmt)
 	if err != nil {
 		return fmt.Errorf("updating software_titles.name: %w", err)
 	}
 
-	dupeIDsStmt := `SELECT
-					s1.id AS id, s1.bundle_identifier AS bundle_identifier FROM software s1
-					JOIN software s2 ON s1.bundle_identifier = s2.bundle_identifier
-						AND s1.version = s2.version
-						AND s1.title_id = s2.title_id
-						AND s1.source = s2.source
-						AND` + " s1.`release` = s2.`release` " +
-		`AND s1.arch = s2.arch
-						AND s1.vendor = s2.vendor
-						AND s1.browser = s2.browser
-						AND s1.extension_id = s2.extension_id
-				WHERE
-					s1.source = 'apps'
+	dupeIDsStmt := `SELECT GROUP_CONCAT(id) AS ids, MD5(
+			-- simulate new hash
+			CONCAT_WS(CHAR(0),
+				version,
+				source,
+				bundle_identifier,
+				` + "`release`" + `,
+				arch,
+				vendor,
+				browser,
+				extension_id
+			)
+		) AS new_checksum
+				FROM software
+				WHERE source = 'apps' AND bundle_identifier IS NOT NULL AND bundle_identifier != ''
 				GROUP BY
-					id
-				HAVING
-					COUNT(*) > 1`
+					version, source, bundle_identifier,` + "`release`" + `, arch, vendor, browser, extension_id
+				HAVING COUNT(*) > 1`
 
 	txx := sqlx.Tx{Tx: tx, Mapper: reflectx.NewMapperFunc("db", sqlx.NameMapper)}
-	selectedIDs := make(map[string]uint)
-	excludedIDs := make(map[string][]uint)
-	var softwareIDs []struct {
-		ID               uint   `db:"id"`
-		BundleIdentifier string `db:"bundle_identifier"`
+	selectedIDs := make(map[string]uint64)
+	idsToMergeByNewChecksum := make(map[string][]uint64)
+	var softwareGroups []struct {
+		IDs         string `db:"ids"`
+		NewChecksum string `db:"new_checksum"`
 	}
-	if err := txx.Select(&softwareIDs, dupeIDsStmt); err != nil {
+	if err := txx.Select(&softwareGroups, dupeIDsStmt); err != nil {
 		if !errors.Is(err, sql.ErrNoRows) {
 			return fmt.Errorf("selecting duplicate software rows: %w", err)
 		}
 	}
 
-	for _, s := range softwareIDs {
-		if _, ok := selectedIDs[s.BundleIdentifier]; !ok {
-			selectedIDs[s.BundleIdentifier] = s.ID
-			continue
-		}
+	for _, s := range softwareGroups {
+		for _, idStr := range strings.Split(s.IDs, ",") {
+			id, err := strconv.ParseUint(idStr, 10, 64)
+			if err != nil {
+				return fmt.Errorf("building duplicate IDs list %q: %w", idStr, err)
+			}
+
+			if _, ok := selectedIDs[s.NewChecksum]; !ok {
+				selectedIDs[s.NewChecksum] = id
+				continue
+			}
 
-		excludedIDs[s.BundleIdentifier] = append(excludedIDs[s.BundleIdentifier], s.ID)
+			idsToMergeByNewChecksum[s.NewChecksum] = append(idsToMergeByNewChecksum[s.NewChecksum], id)
+		}
```

Comment:
Reading this back:
* `softwareGroups` contains a checksum hash and a comma-separated list of software IDs that match that checksum (from `dupeIDsStmt`)
* `selectedIDs` maps each checksum to a single ID that we'll use as the canonical software item for that checksum - after we're all done, only software with IDs in `selectedIDs` will remain in the DB
* `idsToMergeByNewChecksum` maps each checksum to an array of IDs to "merge" into the corresponding ID in `selectedIDs`

We iterate through `softwareGroups`, and for each ID that matches a checksum:
* If this is the first ID we see matching that checksum, record it in `selectedIds[<checksum>]`.  
* Otherwise, add it to the array in `idsToMergeByNewChecksum[<checksum>]`.